pkg/blockbuilder/scheduler/jobs.go

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package scheduler
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+)
+
+var (
+	errNoJobAvailable = errors.New("no job available")
+	errJobNotFound    = errors.New("job not found")
+	errJobNotAssigned = errors.New("job not assigned to worker")
+)
+
+type jobQueue struct {
+	leaseTime time.Duration
+	logger    log.Logger
+
+	mu         sync.Mutex
+	jobs       map[string]*job
+	unassigned jobHeap
+}
+
+func newJobQueue(leaseTime time.Duration, logger log.Logger) *jobQueue {
+	return &jobQueue{
+		leaseTime: leaseTime,
+		logger:    logger,
+
+		jobs: make(map[string]*job),
+	}
+}
+
+// assign assigns the highest-priority unassigned job to the given worker.
+func (s *jobQueue) assign(workerID string) (string, jobSpec, error) {
+	if workerID == "" {
+		return "", jobSpec{}, errors.New("workerID cannot not be empty")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.unassigned.Len() == 0 {
+		return "", jobSpec{}, errNoJobAvailable
+	}
+
+	j := heap.Pop(&s.unassigned).(*job)
+	j.assignee = workerID
+	j.leaseExpiry = time.Now().Add(s.leaseTime)
+	return j.id, j.spec, nil
+}
+
+// addOrUpdate adds a new job or updates an existing job with the given spec.
+func (s *jobQueue) addOrUpdate(id string, spec jobSpec) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if j, ok := s.jobs[id]; ok {
+		// We can only update an unassigned job.
+		if j.assignee == "" {
+			j.spec = spec
+		}
+	} else {
+		j = &job{
+			id:          id,
+			assignee:    "",
+			leaseExpiry: time.Now().Add(s.leaseTime),
+			failCount:   0,
+			spec:        spec,
+		}
+		s.jobs[id] = j
+		heap.Push(&s.unassigned, j)
+	}
+}
+
+// renewLease renews the lease of the job with the given ID for the given
+// worker.
+func (s *jobQueue) renewLease(jobID, workerID string) error {
+	if jobID == "" {
+		return errors.New("jobID cannot be empty")
+	}
+	if workerID == "" {
+		return errors.New("workerID cannot be empty")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	j, ok := s.jobs[jobID]
+	if !ok {
+		return errJobNotFound
+	}
+	if j.assignee != workerID {
+		return errJobNotAssigned
+	}
+
+	j.leaseExpiry = time.Now().Add(s.leaseTime)
+	return nil
+}
+
+// completeJob completes the job with the given ID for the given worker,
+// removing it from the jobQueue.
+func (s *jobQueue) completeJob(jobID, workerID string) error {
+	if jobID == "" {
+		return errors.New("jobID cannot be empty")
+	}
+	if workerID == "" {
+		return errors.New("workerID cannot be empty")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	j, ok := s.jobs[jobID]
+	if !ok {
+		return errJobNotFound
+	}
+	if j.assignee != workerID {
+		return errJobNotAssigned
+	}
```

Comment:
I have one scenario in mind. Job J was assigned to BB1. Lease expires due to communication issue and job gets assigned to BB2 now, but BB1 is still doing the work. Now BB1 and BB2 completes the job. So I guess in that case we ignore the job completion of BB1 and only consider BB2. Right?

Small optimisation: when BBs are updating their job status, scheduler can verify if that job is assigned to that BB, and if not, ask BB to stop that job. Because scheduler will not consider their work for a "complete" status.

Scenarios like this might also affect our in-memory maintenance of committed offset and can drift from what's in kafka. So we should be careful when doing that, i.e. not get the commit offset down, and only re-commit if it is higher than previous commit. Basically scheduler syncs with kafka and keeps the higher among scheduler and kafka and commits where necessary.

Reply:
Yes on all counts. 

> I have one scenario in mind. Job J was assigned to BB1. Lease expires due to communication issue and job gets assigned to BB2 now, but BB1 is still doing the work. Now BB1 and BB2 completes the job. So I guess in that case we ignore the job completion of BB1 and only consider BB2. Right?

That is correct. This is even made tricky in the startup observation code because we're learning from the workers about the jobs. I'm adding an assignment generation number so we can decide the winner.

> Small optimisation: when BBs are updating their job status, scheduler can verify if that job is assigned to that BB, and if not, ask BB to stop that job. Because scheduler will not consider their work for a "complete" status.
> 
> Scenarios like this might also affect our in-memory maintenance of committed offset and can drift from what's in kafka. So we should be careful when doing that, i.e. not get the commit offset down, and only re-commit if it is higher than previous commit. Basically scheduler syncs with kafka and keeps the higher among scheduler and kafka and commits where necessary.